src/core/public/context/context.ts

```diff
@@ -0,0 +1,281 @@
+ * @returns The context value associated with this key. May also return a Promise.
+ *
+ * @public
+ */
+export type ContextProvider<
+  TContext extends {},
+  TContextName extends keyof TContext,
+  TProviderParameters extends any[] = []
+> = (
+  context: Partial<TContext>,
+  ...rest: TProviderParameters
+) => Promise<TContext[TContextName]> | TContext[TContextName];
+
+/**
+ * A function registered by a plugin to perform some action.
+ *
+ * @remarks
+ * A new `TContext` will be built for each handler before invoking.
+ *
+ * @public
+ */
+export type Handler<TContext extends {}, TReturn, THandlerParameters extends any[] = []> = (
+  context: TContext,
+  ...rest: THandlerParameters
+) => TReturn;
+
+type Promisify<T> = T extends Promise<infer U> ? Promise<U> : Promise<T>;
+
+/**
+ * An object that handles registration of context providers and configuring handlers with context.
+ *
+ * @remarks
+ * A {@link ContextContainer} can be used by any Core service or plugin (known as the "service owner") which wishes to
+ * expose APIs in a handler function. The container object will manage registering context providers and configuring a
+ * handler with all of the contexts that should be exposed to the handler's plugin. This is dependent on the
+ * dependencies that the handler's plugin declares.
+ *
+ * Contexts providers are executed in the order they were registered. Each provider gets access to context values
+ * provided by any plugins that it depends on.
+ *
+ * In order to configure a handler with context, you must call the {@link ContextContainer.createHandler} function and
+ * use the returned handler which will automatically build a context object when called.
+ *
+ * When registering context or creating handlers, the _calling plugin's id_ must be provided. Note this should NOT be
+ * the context service owner, but the plugin that is actually registering the context or handler.
+ *
+ * ```ts
+ * // GOOD
+ * class MyPlugin {
+ *   private readonly handlers = new Map();
+ *
+ *   setup(core) {
+ *     this.contextContainer = core.context.createContextContainer();
+ *     return {
+ *       registerContext(plugin, contextName, provider) {
+ *         this.contextContainer.registerContext(plugin, contextName, provider);
+ *       },
+ *       registerRoute(plugin, path, handler) {
+ *         this.handlers.set(
+ *           path,
+ *           this.contextContainer.createHandler(plugin, handler)
+ *         );
+ *       }
+ *     }
+ *   }
+ * }
+ *
+ * // BAD
+ * class MyPlugin {
+ *   private readonly handlers = new Map();
+ *
+ *   setup(core) {
+ *     this.contextContainer = core.context.createContextContainer();
+ *     return {
+ *       registerContext(plugin, contextName, provider) {
+ *         // This would leak this context to all handlers rather tha only plugins that depend on the calling plugin.
+ *         this.contextContainer.registerContext('my_plugin', contextName, provider);
```

Comment:
No, because this id has to be the same as the id of the plugin specified in their `kibana.json` file. The plugin service on the server will refuse to start if there are two plugins with the same id.

Reply:
Alternatively, we could ask plugins to pass some unique symbol that Core creates and passes to plugins via the PluginInitializerContext. Or maybe that's what you mean :)

Reply:
> No, because this id has to be the same as the id of the plugin specified in their kibana.json file.

yeah, but I don't see this logic in PR

>Alternatively, we could ask plugins to pass some unique symbol that Core creates and passes to plugins via the PluginInitializerContext.

yes, that was one of my proposals.

Reply:
> No, because this id has to be the same as the id of the plugin specified in their kibana.json file

But how we can check it? I don't see this logic in the PR

>Alternatively, we could ask plugins to pass some unique symbol that Core creates and passes to plugins via the PluginInitializerContext. 

yes, that was of my proposals

Reply:
Updated to use a symbol provided by the `PluginInitializerContext`